aqueductcore/backend/routers/files.py

```
@@ -202,3 +207,93 @@ async def upload_experiment_file(
+            detail="File list can not be empty."
+        )
+
+    try:
+        for file_name in file_list:
+            pathvalidate.validate_filename(file_name)
+
+        experiment = await get_experiment_by_uuid(
+            user_info=context.user_info,
+            db_session=context.db_session,
+            experiment_uuid=experiment_uuid,
+        )
+
+        if UserScope.EXPERIMENT_DELETE_ALL not in context.user_info.scopes:
+            if UserScope.EXPERIMENT_DELETE_OWN not in context.user_info.scopes:
+                raise AQDPermission(
+                    "The user doesn't have permission to delete files from experiment"
+                )
+
+            if experiment.created_by != context.user_info.uuid:
+                raise AQDDBExperimentNonExisting(
+                    "The user doesn't have permission to delete files from this experiment."
+                )
+
+        experiment_dir = build_experiment_dir_absolute_path(
+            str(settings.experiments_dir_path), experiment_uuid
+        )
+
+        if not os.path.exists(experiment_dir):
+            raise HTTPException(
+                status_code=status.HTTP_400_BAD_REQUEST,
+                detail="No files were found for this experiment",
+            )
+
+        invalid_files = []
+        for file_name in file_list:
+            abs_file_path = os.path.join(experiment_dir, file_name)
+            if not os.path.exists(abs_file_path) or not os.path.isfile(abs_file_path):
+                invalid_files.append(file_name)
+
+        if invalid_files:
+            raise HTTPException(
+                status_code=status.HTTP_400_BAD_REQUEST,
+                detail=f"File(s) not found - {format_list_human_readable(invalid_files)}"
+            )
+
+        dest_file_path = os.path.join(experiment_dir, file_name)
+        if os.path.exists(dest_file_path) and os.path.isfile(dest_file_path):
+            os.remove(dest_file_path)
```

Am I right this is a second-time validation for the case of parallel competing process doing the same? I'd probably let it fail without re-check.

---

Right, there's no need to validate during the delete operation. Just the initial validation is fine.